proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -88,14 +139,187 @@ lemma decodeSchedControlInvocation_wf:
    \<lbrace>valid_sc_ctrl_inv'\<rbrace>, -"
   sorry
 
+lemma decodeSchedcontext_Bind_corres:
+  "list_all2 cap_relation excaps excaps'
+   \<Longrightarrow> corres (ser \<oplus> sc_inv_rel)
+         (invs and valid_sched and sc_at sc_ptr and (\<lambda>s. \<forall>x\<in>set excaps. s \<turnstile> x))
+         (invs' and (\<lambda>s. \<forall>x\<in>set excaps'. valid_cap' x s))
+         (decode_sched_context_bind sc_ptr excaps)
+         (decodeSchedContext_Bind sc_ptr excaps')"
+  apply (clarsimp simp: decode_sched_context_bind_def decodeSchedContext_Bind_def)
+  apply (cases excaps; clarsimp)
+  apply (rename_tac cap list)
+  apply (cases excaps'; clarsimp)
+  apply (rule corres_splitEE')
+     apply (corressimp corres: get_sc_corres)
+     apply (fastforce intro: sc_at'_cross_rel[unfolded cross_rel_def, rule_format])
+    apply (rule liftE_validE[THEN iffD2, OF get_sched_context_sp])
+   apply (rule liftE_validE[THEN iffD2, OF get_sc_sp'])
+  apply (rule corres_splitEE_skip; (solves wpsimp)?)
+   apply (corressimp simp: sc_relation_def)
+  apply (case_tac cap; clarsimp)
+   apply (clarsimp simp: bindE_assoc)
+   apply (rule corres_splitEE'[where r'="(=)"]; (solves wpsimp)?)
+    apply (corressimp corres: get_ntfn_corres
+                        simp: get_sk_obj_ref_def ntfn_relation_def valid_cap_def valid_cap'_def
+                          wp: hoare_vcg_all_lift)
+   apply (rule corres_splitEE_skip; (solves wpsimp)?)
+    apply (corressimp corres: get_ntfn_corres
+                        simp: get_sk_obj_ref_def sc_relation_def)
+   apply (clarsimp simp: returnOk_def)
+  apply (clarsimp simp: bindE_assoc get_tcb_obj_ref_def)
+  apply (rule corres_splitEE'[where r'="(=)"])
+     apply (subst corres_liftE_rel_sum)
+     apply (rule corres_guard_imp)
+       apply (rule threadget_corres)
+       apply (clarsimp simp: tcb_relation_def)
+      apply (clarsimp simp: valid_cap_def)
+     apply (clarsimp simp: valid_cap'_def)
+    apply (rule liftE_validE[THEN iffD2, OF thread_get_sp])
+   apply (rule liftE_validE[THEN iffD2, OF threadGet_sp])
+  apply (rule corres_splitEE_skip; (solves \<open>wpsimp simp: valid_cap'_def obj_at'_def\<close>)?)
+   apply (corressimp corres: get_ntfn_corres
+                       simp: get_sk_obj_ref_def sc_relation_def)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split_eqrE)
+       apply (rule corres_splitEE)
+          apply (rule whenE_throwError_corres)
+            apply simp
+           apply simp
+          apply (clarsimp simp: returnOk_def)
+         apply (subst corres_liftE_rel_sum)
+         apply (rule corres_rel_imp)
```

Comment:
nitpick: instead of two rules and simp, consider `apply (rule xxx[OF yyy], simp)` or similar.